commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java

```diff
@@ -0,0 +1,136 @@
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.scanner.ScannerException;
+import picocli.CommandLine;
+import picocli.CommandLine.IDefaultValueProvider;
+import picocli.CommandLine.Model.ArgSpec;
+import picocli.CommandLine.Model.CommandSpec;
+import picocli.CommandLine.Model.OptionSpec;
+import picocli.CommandLine.ParameterException;
+
+/** Yaml Configuration which is specifically written for Eth2SignerCommand. */
+public class YamlConfigFileDefaultProvider implements IDefaultValueProvider {
+
+  private final CommandLine commandLine;
+  private final File configFile;
+  private Map<String, Object> result;
+
+  public YamlConfigFileDefaultProvider(final CommandLine commandLine, final File configFile) {
+    this.commandLine = commandLine;
+    this.configFile = configFile;
+  }
+
+  @Override
+  public String defaultValue(final ArgSpec argSpec) {
+    loadConfigurationFromFile();
+
+    // only options can be used in config because a name is needed for the key
+    // so we skip default for positional params
+    return argSpec.isOption() ? getConfigurationValue(((OptionSpec) argSpec)) : null;
+  }
+
+  private void loadConfigurationFromFile() {
+    if (result != null) {
+      return;
+    }
+
+    try {
+      final String configYaml = Files.readString(configFile.toPath());
+      final Map<String, Object> result = new Yaml().load(configYaml);
+      checkConfigurationValidity(result == null || result.isEmpty());
+      checkUnknownOptions(result);
+      this.result = result;
+    } catch (final FileNotFoundException | NoSuchFileException e) {
+      throw new ParameterException(
+          commandLine, "Unable to read yaml configuration. File not found: " + configFile, e);
+    } catch (final IOException e) {
+      throw new ParameterException(
+          commandLine,
+          "Unexpected IO error while reading yaml configuration file ["
+              + configFile
+              + "]: "
+              + e.getMessage(),
+          e);
+    } catch (final ScannerException e) {
+      throw new ParameterException(
+          commandLine,
+          "Unable to read yaml configuration. Invalid yaml file ["
+              + configFile
+              + "]: "
+              + e.getMessage(),
+          e);
+    }
+  }
+
+  private void checkUnknownOptions(final Map<String, Object> result) {
+    final CommandSpec commandSpec = commandLine.getCommandSpec();
+
+    final Set<String> unknownOptionsList =
+        result.keySet().stream()
+            .filter(option -> !commandSpec.optionsMap().containsKey("--" + option))
+            .collect(Collectors.toSet());
+
+    if (!unknownOptionsList.isEmpty()) {
+      final String options = unknownOptionsList.size() > 1 ? "options" : "option";
+      final String csvUnknownOptions =
+          unknownOptionsList.stream().collect(Collectors.joining(", "));
```

Comment:
can be replaced with String.join(", ", unknownOptionsList);

Reply:
👍 Updated test cases as well.